README.md
# Todo_app

A todo app built with flutter and firebase

Using SDKMAN & FVM for configure JDK & SDK Flutter

JDK version 8.0.362 (1.8)
SDK Flutter version 2.10.5

<p float= "left">
  <img src = "screenshots/first.png" width = "250" />
  <img src = "screenshots/second.png" width = "250" />
  <img src = "screenshots/third.png" width = "250" />
  <img src = "screenshots/fourth.png" width = "250" />
</p>
<a href = "https://www.uplabs.com/posts/to-do-list-app-concept-cb2ae973-4d2c-4230-b739-132a2b42c521">Design Credit</a>


changelog.txt
Changelog

- Release

lib/main.dart
import 'dart:developer';

import 'package:flutter/material.dart';
import 'package:flutter_bloc/flutter_bloc.dart';
import 'package:onesignal_flutter/onesignal_flutter.dart';

// import 'package:flutter_statusbarcolor/flutter_statusbarcolor.dart';

import './src/injection_container.dart' as di;
import 'src/bloc_delegate.dart';
import 'src/features/todo_manager_features/presentation/bloc/add_list_bloc/add_list_bloc.dart'
    as addListBloc;
import 'src/features/todo_manager_features/presentation/bloc/add_quick_note_bloc/add_quick_note_bloc.dart'
    as addNoteBloc;
import 'src/features/todo_manager_features/presentation/bloc/auth_bloc/auth_bloc.dart';
import 'src/features/todo_manager_features/presentation/bloc/dashboard_bloc/dashboard_bloc.dart';
import 'src/features/todo_manager_features/presentation/bloc/login_bloc/login_bloc.dart';
import 'src/features/todo_manager_features/presentation/pages/Dashboard.dart';
import 'src/features/todo_manager_features/presentation/pages/Login.dart';
import 'src/features/todo_manager_features/presentation/pages/add_new_list.dart';
import 'src/features/todo_manager_features/presentation/pages/add_quick_note.dart';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await di.init();
  BlocSupervisor.delegate = MyBlocDelegate();
  await di.sl.allReady();
  // await FlutterStatusbarcolor.setStatusBarColor(Colors.white.withOpacity(0.90));
  // FlutterStatusbarcolor.setStatusBarWhiteForeground(false);
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  AuthBloc _authBloc;

  // CHANGE THIS parameter to true if you want to test GDPR privacy consent
  bool _requireConsent = true;

  @override
  void initState() {
    super.initState();
    _authBloc = di.sl<AuthBloc>();
    _authBloc.add(AppStarted());

    initPlatformState();
    OneSignal.shared.consentGranted(true);
  }

  @override
  void dispose() {
    _authBloc.close();
    super.dispose();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    if (!mounted) return;

    OneSignal.shared.setLogLevel(OSLogLevel.verbose, OSLogLevel.none);

    OneSignal.shared.setRequiresUserPrivacyConsent(_requireConsent);

    OneSignal.shared
        .setNotificationOpenedHandler((OSNotificationOpenedResult result) {
      print('NOTIFICATION OPENED HANDLER CALLED WITH: ${result}');
      this.setState(() {
        final _debugLabelString =
            "Opened notification: \n${result.notification.jsonRepresentation().replaceAll("\\n", "\n")}";
        log('message :: _debugLabelString :: $_debugLabelString');
      });
    });

    OneSignal.shared.setNotificationWillShowInForegroundHandler(
        (OSNotificationReceivedEvent event) {
      print('FOREGROUND HANDLER CALLED WITH: ${event}');

      /// Display Notification, send null to not display
      event.complete(null);

      this.setState(() {
        final _debugLabelString =
            "Notification received in foreground notification: \n${event.notification.jsonRepresentation().replaceAll("\\n", "\n")}";
        log('message :: _debugLabelString :: $_debugLabelString');
      });
    });

    OneSignal.shared
        .setInAppMessageClickedHandler((OSInAppMessageAction action) {
      this.setState(() {
        final _debugLabelString =
            "In App Message Clicked: \n${action.jsonRepresentation().replaceAll("\\n", "\n")}";
        log('message :: _debugLabelString :: $_debugLabelString');
      });
    });

    OneSignal.shared
        .setSubscriptionObserver((OSSubscriptionStateChanges changes) {
      print("SUBSCRIPTION STATE CHANGED: ${changes.jsonRepresentation()}");
    });

    OneSignal.shared.setPermissionObserver((OSPermissionStateChanges changes) {
      print("PERMISSION STATE CHANGED: ${changes.jsonRepresentation()}");
    });

    OneSignal.shared.setEmailSubscriptionObserver(
        (OSEmailSubscriptionStateChanges changes) {
      print("EMAIL SUBSCRIPTION STATE CHANGED ${changes.jsonRepresentation()}");
    });

    OneSignal.shared
        .setSMSSubscriptionObserver((OSSMSSubscriptionStateChanges changes) {
      print("SMS SUBSCRIPTION STATE CHANGED ${changes.jsonRepresentation()}");
    });

    OneSignal.shared.setOnWillDisplayInAppMessageHandler((message) {
      print("ON WILL DISPLAY IN APP MESSAGE ${message.messageId}");
    });

    OneSignal.shared.setOnDidDisplayInAppMessageHandler((message) {
      print("ON DID DISPLAY IN APP MESSAGE ${message.messageId}");
    });

    OneSignal.shared.setOnWillDismissInAppMessageHandler((message) {
      print("ON WILL DISMISS IN APP MESSAGE ${message.messageId}");
    });

    OneSignal.shared.setOnDidDismissInAppMessageHandler((message) {
      print("ON DID DISMISS IN APP MESSAGE ${message.messageId}");
    });

    // NOTE: Replace with your own app ID from https://www.onesignal.com
    await OneSignal.shared.setAppId("85c99d38-bdca-4d16-91a9-fa6d89e311a7");

    // iOS-only method to open launch URLs in Safari when set to false
    OneSignal.shared.setLaunchURLsInApp(false);

    bool requiresConsent = await OneSignal.shared.requiresUserPrivacyConsent();

    this.setState(() {
      // _enableConsentButton = requiresConsent;
    });

    // Some examples of how to use In App Messaging public methods with OneSignal SDK
    oneSignalInAppMessagingTriggerExamples();

    OneSignal.shared.disablePush(false);

    // Some examples of how to use Outcome Events public methods with OneSignal SDK
    oneSignalOutcomeEventsExamples();

    bool userProvidedPrivacyConsent =
        await OneSignal.shared.userProvidedPrivacyConsent();
    print("USER PROVIDED PRIVACY CONSENT: $userProvidedPrivacyConsent");
  }

  oneSignalInAppMessagingTriggerExamples() async {
    /// Example addTrigger call for IAM
    /// This will add 1 trigger so if there are any IAM satisfying it, it
    /// will be shown to the user
    OneSignal.shared.addTrigger("trigger_1", "one");

    /// Example addTriggers call for IAM
    /// This will add 2 triggers so if there are any IAM satisfying these, they
    /// will be shown to the user
    Map<String, Object> triggers = new Map<String, Object>();
    triggers["trigger_2"] = "two";
    triggers["trigger_3"] = "three";
    OneSignal.shared.addTriggers(triggers);

    // Removes a trigger by its key so if any future IAM are pulled with
    // these triggers they will not be shown until the trigger is added back
    OneSignal.shared.removeTriggerForKey("trigger_2");

    // Get the value for a trigger by its key
    final triggerValue =
        await OneSignal.shared.getTriggerValueForKey("trigger_3");
    print("'trigger_3' key trigger value: ${triggerValue?.toString()}");

    // Create a list and bulk remove triggers based on keys supplied
    List<String> keys = ["trigger_1", "trigger_3"];
    OneSignal.shared.removeTriggersForKeys(keys);

    // Toggle pausing (displaying or not) of IAMs
    OneSignal.shared.pauseInAppMessages(false);
  }

  oneSignalOutcomeEventsExamples() async {
    // Await example for sending outcomes
    outcomeAwaitExample();

    // Send a normal outcome and get a reply with the name of the outcome
    OneSignal.shared.sendOutcome("normal_1");
    OneSignal.shared.sendOutcome("normal_2").then((outcomeEvent) {
      print(outcomeEvent.jsonRepresentation());
    });

    // Send a unique outcome and get a reply with the name of the outcome
    OneSignal.shared.sendUniqueOutcome("unique_1");
    OneSignal.shared.sendUniqueOutcome("unique_2").then((outcomeEvent) {
      print(outcomeEvent.jsonRepresentation());
    });

    // Send an outcome with a value and get a reply with the name of the outcome
    OneSignal.shared.sendOutcomeWithValue("value_1", 3.2);
    OneSignal.shared.sendOutcomeWithValue("value_2", 3.9).then((outcomeEvent) {
      print(outcomeEvent.jsonRepresentation());
    });
  }

  Future<void> outcomeAwaitExample() async {
    var outcomeEvent = await OneSignal.shared.sendOutcome("await_normal_1");
    print(outcomeEvent.jsonRepresentation());
  }

  @override
  Widget build(BuildContext context) {
    return BlocProvider.value(
      value: _authBloc,
      child: MaterialApp(
        debugShowCheckedModeBanner: false,
        title: 'Nxt App',
        theme: ThemeData(
          primaryColor: Color(0xff657AFF),
          accentColor: Color(0xff4F5578),
          primarySwatch: Colors.blue,
        ),
        home: BlocBuilder(
            bloc: _authBloc,
            builder: (BuildContext context, AuthState authState) {
              if (authState is Uninitialised) {
                return BlocProvider<LoginBloc>(
                    create: (context) {
                      return di.sl<LoginBloc>();
                    },
                    child: LoginPage());
              } else if (authState is AuthAuthenticated) {
                return FutureBuilder(
                    future: di.sl.allReady(),
                    builder: (context, snapshot) {
                      if (snapshot.hasData) {
                        return BlocProvider<DashboardBloc>(
                          create: (context) {
                            return di.sl<DashboardBloc>();
                          },
                          child: DashboardScreen(),
                        );
                      } else {
                        return Scaffold(
                          body: Center(
                            child: Container(
                              height: 40,
                              width: 40,
                              child: CircularProgressIndicator(),
                            ),
                          ),
                        );
                      }
                    });
              } else {
                return BlocProvider<LoginBloc>(
                    create: (context) {
                      return di.sl<LoginBloc>();
                    },
                    child: LoginPage());
              }
            }),
        routes: {
          "/dashboard": (context) => BlocProvider<DashboardBloc>(
                create: (context) {
                  return di.sl<DashboardBloc>();
                },
                child: DashboardScreen(),
              ),
          "/addQuickNote": (context) =>
              BlocProvider<addNoteBloc.AddQuickNoteBloc>(
                  create: (context) {
                    return di.sl<addNoteBloc.AddQuickNoteBloc>();
                  },
                  child: AddQuickNote()),
          "/addList": (context) => BlocProvider<addListBloc.AddListBloc>(
              create: (context) {
                return di.sl<addListBloc.AddListBloc>();
              },
              child: AddList()),
        },
      ),
    );
  }
}


lib/src/features/todo_manager_features/presentation/pages/Dashboard.dart
import 'package:flutter/material.dart';
import 'package:flutter_bloc/flutter_bloc.dart';
import 'package:todo_app/src/features/todo_manager_features/presentation/widgets/no_widget.dart';

import '../../../../core/user/user.dart';
import '../../../../injection_container.dart';
import '../../domain/entities/List.dart';
import '../../domain/usecases/dashboard_use_cases.dart';
import '../bloc/auth_bloc/auth_bloc.dart';
import '../bloc/dashboard_bloc/dashboard_bloc.dart';
import '../widgets/all_lists.dart';
import '../widgets/all_quick_notes.dart';
import '../widgets/list_bottom_sheet.dart';
import '../widgets/menu_icon.dart';
import '../widgets/small_button.dart';

class DashboardScreen extends StatelessWidget {
  final GlobalKey<ScaffoldState> _scaffoldKey = GlobalKey();

  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Scaffold(
        backgroundColor: Colors.white,
        // resizeToAvoidBottomPadding: false,
        key: _scaffoldKey,

        body: Dashboard(
            bloc: BlocProvider.of<DashboardBloc>(context),
            scaffoldKey: _scaffoldKey),
        endDrawer: Container(
          padding:
              EdgeInsets.only(left: 20.0, right: 20.0, top: 40.0, bottom: 20.0),
          width: MediaQuery.of(context).size.width,
          color: Colors.white,
          child: Drawer(
            elevation: 0.0,
            child: Container(
              color: Colors.white,
              width: double.infinity,
              height: double.infinity,
              child: Row(
                mainAxisAlignment: MainAxisAlignment.end,
                children: <Widget>[
                  Column(
                    crossAxisAlignment: CrossAxisAlignment.end,
                    children: <Widget>[
                      SmallButton(
                        color: Color(0xff878CAC),
                        icon: Icon(
                          Icons.clear,
                          color: Colors.white,
                        ),
                        onPressed: () {
                          Navigator.pop(context);
                        },
                      ),
                      SizedBox(height: 30.0),
                      Row(
                        children: <Widget>[
                          Text(
                            "+ Add a quick note",
                            style: TextStyle(
                                color: Theme.of(context).accentColor,
                                fontFamily: "Poppins",
                                fontSize: 20,
                                fontWeight: FontWeight.bold),
                          ),
                          SizedBox(width: 30.0),
                          SmallButton(
                            icon: Transform.rotate(
                              angle: 45.0,
                              child: Icon(
                                Icons.attach_file,
                                color: Colors.white,
                              ),
                            ),
                            onPressed: () async {
                              print(Navigator);
                              //Navigator.pop(context);
                              await Navigator.popAndPushNamed(
                                  context, "/addQuickNote");
                            },
                          )
                        ],
                      ),
                      SizedBox(height: 30.0),
                      Row(
                        children: <Widget>[
                          Text(
                            "+ Add a list",
                            style: TextStyle(
                                color: Theme.of(context).accentColor,
                                fontFamily: "Poppins",
                                fontSize: 20,
                                fontWeight: FontWeight.bold),
                          ),
                          SizedBox(width: 30.0),
                          SmallButton(
                            icon: Icon(
                              Icons.insert_drive_file,
                              color: Colors.white,
                            ),
                            onPressed: () async {
                              await Navigator.popAndPushNamed(
                                  context, "/addList");
                            },
                          )
                        ],
                      ),
                      SizedBox(height: 30.0),
                      Row(
                        children: <Widget>[
                          Text(
                            "Sign out",
                            style: TextStyle(
                                color: Theme.of(context).accentColor,
                                fontFamily: "Poppins",
                                fontSize: 20,
                                fontWeight: FontWeight.bold),
                          ),
                          SizedBox(width: 30.0),
                          SmallButton(
                            color: Color(0xff878CAC),
                            icon: Icon(
                              Icons.exit_to_app,
                              color: Colors.white,
                            ),
                            onPressed: () async {
                              BlocProvider.of<AuthBloc>(context)
                                  .add(LoggedOut());
                              Navigator.of(context).pop();
                            },
                          )
                        ],
                      )
                    ],
                  ),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}

class Dashboard extends StatefulWidget {
  final GlobalKey<ScaffoldState> scaffoldKey;
  final DashboardBloc bloc;
  Dashboard({
    this.bloc,
    this.scaffoldKey,
  });

  @override
  _DashboardState createState() => _DashboardState();
}

class _DashboardState extends State<Dashboard> {
  PersistentBottomSheetController _controller;

  @override
  void initState() {
    super.initState();
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
  }

  @override
  void dispose() {
    super.dispose();
  }

  void _showBottomSheet(ListEntity list) {
    _controller = this.widget.scaffoldKey.currentState.showBottomSheet(
        (context) {
      return AnimatedPadding(
        padding: MediaQuery.of(context).viewInsets,
        duration: const Duration(milliseconds: 100),
        curve: Curves.decelerate,
        child: ListBottomSheet(
          controller: _controller,
          todoList: list,
          closeBottomSheet: () {
            _closeBottomSheet();
          },
        ),
      );
    },
        shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.only(topRight: Radius.circular(50))));
  }

  void _closeBottomSheet() {
    _controller.close();
  }

  @override
  Widget build(BuildContext context) {
    return ListView(
      padding:
          EdgeInsets.only(left: 30.0, right: 30.0, top: 40.0, bottom: 20.0),
      shrinkWrap: true,
      children: <Widget>[
        AppBar(
          leading: MenuIcon(),
          actions: <Widget>[NoWidget()],
          elevation: 0.0,
          backgroundColor: Colors.transparent,
        ),
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: <Widget>[
            Text(
              'Hello ${sl<User>().displayName.toString().split(" ")[0]}',
              style: TextStyle(
                  fontSize: 32.0,
                  fontFamily: "Poppins",
                  fontWeight: FontWeight.w200,
                  color: Theme.of(context).accentColor),
            ),
            SmallButton(
              icon: Icon(
                Icons.add,
                color: Colors.white,
              ),
              onPressed: () {
                this.widget.scaffoldKey.currentState.openEndDrawer();
              },
            )
          ],
        ),
        SizedBox(height: 40.0),
        Row(
          mainAxisAlignment: MainAxisAlignment.start,
          children: <Widget>[
            Text(
              "Quick notes",
              style: TextStyle(
                  fontSize: 22.0,
                  fontWeight: FontWeight.bold,
                  fontFamily: "Poppins",
                  color: Theme.of(context).accentColor),
            )
          ],
        ),
        SizedBox(height: 20.0),
        AllQuickNotes(
            allQuickNotesStream:
                sl<FetchQuickNotes>().call(uid: sl<User>().uid)),
        SizedBox(height: 70.0),
        Row(
          mainAxisAlignment: MainAxisAlignment.start,
          children: <Widget>[
            Text(
              "Lists",
              style: TextStyle(
                  fontSize: 22.0,
                  fontWeight: FontWeight.bold,
                  fontFamily: "Poppins",
                  color: Theme.of(context).accentColor),
            ),
          ],
        ),
        SizedBox(height: 30.0),
        AllLists(
            allListsStream: sl<FetchLists>().call(uid: sl<User>().uid),
            show: this._showBottomSheet)
      ],
    );
  }
}


lib/src/features/todo_manager_features/presentation/bloc/add_list_bloc/add_list_bloc.dart
import 'dart:async';

import 'package:bloc/bloc.dart';
import 'package:equatable/equatable.dart';
import 'package:meta/meta.dart';
import 'package:todo_app/src/core/user/user.dart';
import 'package:todo_app/src/features/todo_manager_features/domain/entities/ListTodo.dart';
import 'package:todo_app/src/features/todo_manager_features/domain/usecases/add_list.dart'
    as usecase;

part 'add_list_event.dart';
part 'add_list_state.dart';

class AddListBloc extends Bloc<AddListEvent, AddListState> {
  final usecase.AddList _addList;
  final User _user;

  AddListBloc({@required usecase.AddList addList, @required User user})
      : assert(addList != null),
        _addList = addList,
        assert(user != null),
        _user = user;
  @override
  AddListState get initialState => AddListInitial();

  @override
  Stream<AddListState> mapEventToState(
    AddListEvent event,
  ) async* {
    if (event is AddList) {
      yield AddingList();
      await _addList(
          uid: _user.uid,
          listTitle: event.listTitle,
          listOfTodos: event.listOfTodos,
          listBackgroundColor: event.listBackgroundColor);
      yield ListAdded();
    }
  }
}


lib/src/features/todo_manager_features/presentation/pages/add_new_list.dart
import 'package:flutter/material.dart';
import 'package:flutter_bloc/flutter_bloc.dart';
import 'package:todo_app/src/features/todo_manager_features/presentation/bloc/add_list_bloc/add_list_bloc.dart'
    as event;

import '../../../../core/constants/Constants.dart';
import '../../domain/entities/ListTodo.dart';
import '../bloc/add_list_bloc/add_list_bloc.dart';
import '../widgets/small_button.dart';

class AddList extends StatefulWidget {
  @override
  _AddListState createState() => _AddListState();
}

class _AddListState extends State<AddList> {
  Color _selectedColor = listOfColorsForColorPicker[2];
  List<ListTodo> _listOfTodos = [];

  final TextEditingController _newTodoController = new TextEditingController();
  final TextEditingController _titleController = new TextEditingController();
  final GlobalKey<FormState> _formKey = new GlobalKey<FormState>();

  @override
  void dispose() {
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return BlocListener<AddListBloc, AddListState>(
      listener: (context, state) {
        if (state is AddingList) {
          if (_formKey.currentState.validate()) {
            _formKey.currentState.save();
            showDialog(
                context: context,
                builder: (context) {
                  return AlertDialog(
                    content: Row(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: <Widget>[
                        Container(
                            width: 40,
                            height: 40,
                            child: CircularProgressIndicator()),
                      ],
                    ),
                  );
                });
          }
        } else if (state is ListAdded) {
          Navigator.pop(context);
          showDialog(
              context: context,
              builder: (context) {
                return AlertDialog(
                  title: Text("New List added"),
                );
              });
          setState(() {
            _titleController.text = "";
            _listOfTodos.clear();
          });
        }
      },
      child: Scaffold(
          backgroundColor: Colors.white,
          body: ListView(children: <Widget>[
            Padding(
              padding:
                  const EdgeInsets.only(top: 20.0, left: 10.0, right: 20.0),
              child: AppBar(
                elevation: 0.0,
                backgroundColor: Colors.transparent,
                leading: Row(
                  mainAxisAlignment: MainAxisAlignment.start,
                  children: <Widget>[
                    IconButton(
                      color: Colors.yellow,
                      padding: EdgeInsets.all(0.0),
                      onPressed: () {
                        Navigator.pop(context);
                      },
                      icon: Icon(Icons.clear,
                          size: 26.0, color: Color(0xff616B77)),
                    ),
                  ],
                ),
              ),
            ),
            Padding(
                padding: EdgeInsets.only(
                    left: 25.0, right: 25.0, top: 10.0, bottom: 20.0),
                child: Column(children: <Widget>[
                  Form(
                    key: _formKey,
                    child: Row(
                      children: <Widget>[
                        Expanded(
                          child: TextFormField(
                            controller: _titleController,
                            validator: (val) {
                              if (val.isEmpty) {
                                return 'Please enter some text';
                              }
                              return null;
                            },
                            style: TextStyle(
                                fontSize: 27.0,
                                fontFamily: "Poppins",
                                color: Theme.of(context).accentColor),
                            decoration: InputDecoration(
                                hintText: "Title of new list...",
                                border: UnderlineInputBorder(
                                    borderSide: BorderSide.none)),
                          ),
                        ),
                        SizedBox(width: 20.0),
                        SmallButton(
                            icon: Icon(
                              Icons.check,
                              color: Colors.white,
                            ),
                            onPressed: () async {
                              if (_formKey.currentState.validate()) {
                                BlocProvider.of<AddListBloc>(context).add(
                                    event.AddList(
                                        listBackgroundColor:
                                            "#${_selectedColor.toString().substring(10, 16)}",
                                        listOfTodos: _listOfTodos,
                                        listTitle: _titleController.text));
                              }
                            })
                      ],
                    ),
                  ),
                  SizedBox(height: 30.0),
                  Row(
                    mainAxisAlignment: MainAxisAlignment.spaceBetween,
                    children: listOfColorsForColorPicker.map((Color color) {
                      return GestureDetector(
                        onTap: () {
                          setState(() {
                            _selectedColor = color;
                          });
                        },
                        child: Container(
                          height: 30,
                          width: 30,
                          decoration: BoxDecoration(
                            borderRadius: BorderRadius.circular(30.0),
                            color: color,
                            border: color == Color(0xffFFFFFF)
                                ? Border.all(
                                    width: 1,
                                    color: Theme.of(context).accentColor)
                                : Border.all(width: 0, color: color),
                          ),
                          child: Center(
                            child: _selectedColor == color
                                ? Container(
                                    height: 10,
                                    width: 10,
                                    decoration: BoxDecoration(
                                        color: Colors.white,
                                        borderRadius:
                                            BorderRadius.circular(10)),
                                  )
                                : SizedBox(width: 0, height: 0),
                          ),
                        ),
                      );
                    }).toList(),
                  ),
                  SizedBox(height: 50.0),
                  Row(
                    mainAxisAlignment: MainAxisAlignment.start,
                    children: <Widget>[
                      Text(
                        "List items",
                        style: TextStyle(
                            fontFamily: "Poppins",
                            fontWeight: FontWeight.bold,
                            fontSize: 21,
                            color: Theme.of(context).accentColor),
                      )
                    ],
                  ),
                  SizedBox(height: 10),
                  _listOfTodos.length >= 1
                      ? Column(
                          children: _listOfTodos.map((ListTodo todo) {
                            return Container(
                              margin: EdgeInsets.only(bottom: 30.0),
                              child: Row(
                                children: <Widget>[
                                  Container(
                                    height: 7,
                                    width: 7,
                                    decoration: BoxDecoration(
                                        borderRadius:
                                            BorderRadius.circular(15.0),
                                        color: _selectedColor),
                                  ),
                                  SizedBox(width: 40.0),
                                  Container(
                                    width: MediaQuery.of(context).size.width *
                                                0.8 >
                                            300
                                        ? 300
                                        : MediaQuery.of(context).size.width,
                                    child: Text(
                                      todo.title,
                                      style: TextStyle(
                                          fontSize: 20.0,
                                          fontFamily: "Poppins",
                                          color: Theme.of(context).accentColor),
                                    ),
                                  )
                                ],
                              ),
                            );
                          }).toList(),
                        )
                      : SizedBox(height: 0, width: 0),
                  SizedBox(height: 10.0),
                  Row(
                    children: <Widget>[
                      GestureDetector(
                        onTap: () {
                          setState(() {
                            _listOfTodos.add(new ListTodo(
                                isDone: false,
                                title: _newTodoController.text,
                                details: ""));
                            print(
                                'Hello todo add ::: ${_listOfTodos[0].title}');
                            _newTodoController.text = "";
                          });
                        },
                        child: Container(
                          height: 25,
                          width: 25,
                          decoration: BoxDecoration(
                              borderRadius: BorderRadius.circular(5.0),
                              color: Theme.of(context).primaryColor),
                          child: Center(
                            child: Icon(
                              Icons.add,
                              color: Colors.white,
                            ),
                          ),
                        ),
                      ),
                      SizedBox(width: 40.0),
                      Expanded(
                        child: TextField(
                          controller: _newTodoController,
                          style: TextStyle(
                              fontSize: 20.0,
                              fontFamily: "Poppins",
                              color: Theme.of(context).accentColor),
                          decoration: InputDecoration(
                              hintText: "Add new note",
                              border: UnderlineInputBorder(
                                  borderSide: BorderSide.none)),
                        ),
                      ),
                    ],
                  )
                ])),
          ])),
    );
  }
}
